File: app/py/cohort_stats/genes.py
"""Genes tab — HGNC-annotated genomic interpretations.

Includes the Gene × Diagnosis explorer (stacked bars / heatmap), which is
recomputed on demand when the user changes a dropdown.
"""

from __future__ import annotations

from .utils import counter, sorted_entries

PATHOGENIC_CLASSES = {"PATHOGENIC", "LIKELY_PATHOGENIC"}


def collect_gene_rows(packets: list) -> list:
    """Flatten every genomic interpretation into one record.

    This is the single source of truth for everything on the Genes tab —
    the stat cards, both tables and the explorer all derive from it.
    """
    rows = []
    for p in packets:
        sex = (p.get("subject") or {}).get("sex") or "UNKNOWN"
        patient_id = (p.get("subject") or {}).get("id") or p.get("id")
        for interp in p.get("interpretations") or []:
            dx = interp.get("diagnosis") or {}
            for gi in dx.get("genomicInterpretations") or []:
                vi = gi.get("variantInterpretation") or {}
                vd = vi.get("variationDescriptor") or {}
                gene = vd.get("geneContext") or gi.get("gene") or {}
                if not gene.get("symbol"):
                    continue
                expressions = vd.get("expressions") or []
                rows.append(
                    {
                        "gene": gene["symbol"],
                        "hgnc_id": gene.get("valueId") or "—",
                        "disease": (dx.get("disease") or {}).get("label") or "—",
                        "disease_id": (dx.get("disease") or {}).get("id") or "—",
                        "acmg": vi.get("acmgPathogenicityClassification") or "—",
                        "status": gi.get("interpretationStatus") or "—",
                        "hgvs": (expressions[0].get("value") if expressions else None) or "—",
                        "allelic": (vd.get("allelicState") or {}).get("label") or "—",
                        "sex": sex,
                        "patient_id": patient_id,
                    }
                )
    return rows


def compute(packets: list) -> dict:
    n = len(packets)
    rows = collect_gene_rows(packets)

    unique_genes = len({r["gene"] for r in rows})
    unique_dx = len({r["disease_id"] for r in rows})
    unique_patients = len({r["patient_id"] for r in rows})
    path_count = sum(1 for r in rows if r["acmg"] in PATHOGENIC_CLASSES)

    gene_counts = counter(r["gene"] for r in rows)
    top20 = sorted_entries(gene_counts)[:20][::-1]

    acmg_counts = counter(r["acmg"] for r in rows)

    # One summary row per gene
    summary_map: dict = {}
    for r in rows:
        g = summary_map.setdefault(
            r["gene"],
            {
                "gene": r["gene"],
                "hgnc_id": r["hgnc_id"],
                "patients": set(),
                "diseases": set(),
                "acmg_counts": {},
                "records": 0,
            },
        )
        g["patients"].add(r["patient_id"])
        g["diseases"].add(r["disease"])
        g["acmg_counts"][r["acmg"]] = g["acmg_counts"].get(r["acmg"], 0) + 1
        g["records"] += 1

    gene_summary = sorted(
        (
            {
                "gene": g["gene"],
                "hgnc_id": g["hgnc_id"],
                "n_patients": len(g["patients"]),
                "pct_patients": (len(g["patients"]) / n * 100) if n else 0.0,
                "n_diseases": len(g["diseases"]),
                "diseases": " · ".join(sorted(g["diseases"])),
                "records": g["records"],
                "top_acmg": sorted_entries(g["acmg_counts"])[0][0] if g["acmg_counts"] else "—",
                "acmg_break": ", ".join(
                    f"{k.replace('_', ' ')}:{v}" for k, v in sorted_entries(g["acmg_counts"])
                ),
            }
            for g in summary_map.values()
        ),
        key=lambda g: g["n_patients"],
        reverse=True,
    )

    return {
        "stats": {
            "unique_genes": unique_genes,
            "unique_patients": unique_patients,
            "records": len(rows),
            "linked_dx": unique_dx,
            "path_count": path_count,
        },
        "top20": {"labels": [e[0] for e in top20], "counts": [e[1] for e in top20]},
        "acmg_counts": acmg_counts,
        "gene_summary": gene_summary,
        "rows": rows,
    }


# ── Gene × Diagnosis explorer ──────────────────────────────────────────────

def _count(rows: list, metric: str) -> int:
    if metric == "patients":
        return len({r["patient_id"] for r in rows})
    return len(rows)


def explorer(packets: list, view: str = "gene-per-dx", top_n: int = 15,
             metric: str = "patients") -> dict:
    """Data for the Gene × Diagnosis explorer.

    view   — "gene-per-dx" | "dx-per-gene" | "heatmap"
    top_n  — 0 means "all"
    metric — "patients" (unique) | "records"
    """
    rows = collect_gene_rows(packets)
    if not rows:
        return {"view": view, "empty": True}

    all_genes = list(dict.fromkeys(r["gene"] for r in rows))
    all_dx = list(dict.fromkeys(r["disease"] for r in rows))

    by_gene: dict = {}
    by_dx: dict = {}
    for r in rows:
        by_gene.setdefault(r["gene"], []).append(r)
        by_dx.setdefault(r["disease"], []).append(r)

    gene_totals = {g: _count(by_gene[g], metric) for g in all_genes}
    dx_totals = {d: _count(by_dx[d], metric) for d in all_dx}

    if view == "heatmap":
        top_genes = [g for g, _ in sorted_entries(gene_totals)[: top_n or len(all_genes)]]
        top_dx = [d for d, _ in sorted_entries(dx_totals)[: top_n or len(all_dx)]]
        matrix = [
            [
                _count([r for r in by_gene[g] if r["disease"] == d], metric)
                for d in top_dx
            ]
            for g in top_genes
        ]
        max_val = max((v for row in matrix for v in row), default=0) or 1
        return {
            "view": view,
            "empty": False,
            "genes": top_genes,
            "diagnoses": top_dx,
            "matrix": matrix,
            "max": max_val,
            "metric": metric,
        }

    if view == "gene-per-dx":
        primary = [d for d, _ in sorted_entries(dx_totals)[: top_n or len(all_dx)]]
        secondary = list(dict.fromkeys(
            r["gene"] for r in rows if r["disease"] in set(primary)
        ))
        secondary.sort(key=lambda s: gene_totals.get(s, 0), reverse=True)
        datasets = [
            {
                "label": s,
                "data": [
                    _count([r for r in by_dx[pk] if r["gene"] == s], metric)
                    for pk in primary
                ],
            }
            for s in secondary
        ]
    else:  # dx-per-gene
        primary = [g for g, _ in sorted_entries(gene_totals)[: top_n or len(all_genes)]]
        secondary = list(dict.fromkeys(
            r["disease"] for r in rows if r["gene"] in set(primary)
        ))
        secondary.sort(key=lambda s: dx_totals.get(s, 0), reverse=True)
        datasets = [
            {
                "label": s,
                "data": [
                    _count([r for r in by_gene[pk] if r["disease"] == s], metric)
                    for pk in primary
                ],
            }
            for s in secondary
        ]

    return {
        "view": view,
        "empty": False,
        "primary_keys": primary,
        "datasets": datasets,
        "metric": metric,
    }

File: app/js/tabs/genes.js
// Genes tab — renders app/py/cohort_stats/genes.py output, including the
// interactive Gene × Diagnosis explorer (recomputed in Python on demand).

import { charts, destroyChart, hbar, donut, barCell, GD_PALETTE, NUM_FONT } from '../core/charts.js';
import { esc } from '../core/format.js';
import { renderTable } from '../core/tables.js';
import { callApi } from '../core/pybridge.js';

const ACMG_COLORS = {
  PATHOGENIC: '#dc2626', LIKELY_PATHOGENIC: '#d97706',
  UNCERTAIN_SIGNIFICANCE: '#94a3b8', LIKELY_BENIGN: '#16a34a', BENIGN: '#2563eb',
};
const ACMG_BADGE = {
  PATHOGENIC: 'red', LIKELY_PATHOGENIC: 'orange', UNCERTAIN_SIGNIFICANCE: 'muted',
  LIKELY_BENIGN: 'green', BENIGN: 'blue',
};
const STATUS_BADGE = { CAUSATIVE: 'green', CONTRIBUTORY: 'orange', CANDIDATE: 'purple', UNCERTAIN: 'muted' };

const badge = (value, map) =>
  `<span class="badge badge-${map[value] || 'muted'}">${esc(String(value).replace(/_/g, ' '))}</span>`;

export default {
  id: 'genes',
  label: 'Genes',
  computeKey: 'genes',

  render(section, g) {
    section.innerHTML = `
      <div class="section-title">Genes</div>
      <div class="section-subtitle">HGNC-annotated genomic interpretations across the cohort</div>
      <div class="stat-cards" id="gene-stats"></div>
      <div class="grid-2" style="margin-bottom:20px">
        <div class="card">
          <h3>Top 20 Genes by Patient Count</h3>
          <div class="chart-wrap" style="height:300px"><canvas id="chart-genes"></canvas></div>
        </div>
        <div class="card">
          <h3>ACMG Classification</h3>
          <div class="donut-wrap" style="justify-content:center;padding:20px 0">
            <div style="width:160px;height:160px;flex-shrink:0"><canvas id="chart-acmg"></canvas></div>
            <div class="donut-legend" id="acmg-legend"></div>
          </div>
        </div>
      </div>
      <div class="card" style="margin-bottom:20px">
        <h3>Gene Summary
          <span style="font-size:10px;color:var(--muted);text-transform:none;letter-spacing:0;margin-left:8px">one row per gene</span>
        </h3>
        <div id="tbl-gene-summary"></div>
      </div>
      <div class="card" style="margin-bottom:20px">
        <div style="display:flex;align-items:center;justify-content:space-between;flex-wrap:wrap;gap:12px;margin-bottom:16px">
          <div>
            <h3 style="margin-bottom:2px">Gene × Diagnosis Explorer</h3>
            <div style="font-size:11px;color:var(--muted)">Explore how genes and diagnoses relate across the cohort</div>
          </div>
          <div style="display:flex;gap:10px;flex-wrap:wrap;align-items:center">
            <select id="gd-view" class="rows-select">
              <option value="gene-per-dx">Genes within each Diagnosis</option>
              <option value="dx-per-gene">Diagnoses within each Gene</option>
              <option value="heatmap">Co-occurrence Heatmap</option>
            </select>
            <select id="gd-top" class="rows-select">
              <option value="15">Top 15</option>
              <option value="25">Top 25</option>
              <option value="50">Top 50</option>
              <option value="0">All</option>
            </select>
            <select id="gd-metric" class="rows-select">
              <option value="patients">by patients</option>
              <option value="records">by records</option>
            </select>
          </div>
        </div>
        <div id="gd-chart-wrap" style="position:relative;overflow-x:auto">
          <div id="gd-placeholder" style="text-align:center;padding:60px;color:var(--muted);font-size:12px">Loading…</div>
          <canvas id="chart-gene-dx" style="display:none"></canvas>
          <div id="gd-heatmap" style="display:none;overflow-x:auto"></div>
        </div>
        <div id="gd-legend" style="display:flex;flex-wrap:wrap;gap:8px;margin-top:12px;font-size:11px"></div>
      </div>
      <div class="card">
        <h3>Gene × Variant Table</h3>
        <div id="tbl-genes"></div>
      </div>`;

    // ── Stat cards ──
    const s = g.stats;
    document.getElementById('gene-stats').innerHTML = `
      <div class="stat-card"><div class="val" style="color:var(--accent)">${s.unique_genes.toLocaleString()}</div><div class="lbl">Unique Genes</div></div>
      <div class="stat-card"><div class="val" style="color:var(--green)">${s.unique_patients.toLocaleString()}</div><div class="lbl">Patients with Variants</div></div>
      <div class="stat-card"><div class="val" style="color:var(--purple)">${s.records.toLocaleString()}</div><div class="lbl">Total Records</div></div>
      <div class="stat-card"><div class="val" style="color:var(--orange)">${s.linked_dx.toLocaleString()}</div><div class="lbl">Linked Diagnoses</div></div>
      <div class="stat-card"><div class="val" style="color:var(--red)">${s.path_count.toLocaleString()}</div><div class="lbl">Path / Likely Path</div></div>`;

    hbar('genes', 'chart-genes', g.top20.labels, g.top20.counts, '#a78bfa', 'records');

    // ACMG donut with compact legend labels
    donut('acmg', 'chart-acmg', 'acmg-legend', g.acmg_counts, ACMG_COLORS, null);
    document.getElementById('acmg-legend').innerHTML = Object.keys(g.acmg_counts).map(l =>
      `<div class="legend-item"><div class="legend-dot" style="background:${ACMG_COLORS[l] || '#94a3b8'}"></div><span style="font-size:10px">${esc(l.replace(/_/g, ' '))}</span><span class="legend-val">${g.acmg_counts[l]}</span></div>`).join('');

    // ── Gene summary table ──
    const sumCols = [
      { key: 'gene', label: 'Gene', render: r => `<strong style="color:var(--accent)">${esc(r.gene)}</strong>` },
      { key: 'hgnc_id', label: 'HGNC', render: r => `<span style="color:var(--muted);font-size:11px">${esc(r.hgnc_id)}</span>` },
      { key: 'n_patients', label: 'Patients', render: r => barCell(r.n_patients, r.pct_patients, 'var(--purple)') },
      { key: 'records', label: 'Records', render: r => r.records },
      { key: 'n_diseases', label: '# Diagnoses', render: r => r.n_diseases },
      { key: 'diseases', label: 'Diagnoses', render: r => `<span style="font-size:11px">${esc(r.diseases)}</span>` },
      { key: 'top_acmg', label: 'Top ACMG', render: r => badge(r.top_acmg, ACMG_BADGE) },
      { key: 'acmg_break', label: 'ACMG Detail', render: r => `<span style="font-size:10px;color:var(--muted)">${esc(r.acmg_break)}</span>` },
    ];
    renderTable('tbl-gene-summary', g.gene_summary, sumCols,
      { defaultSort: 'n_patients', defaultDir: 'desc', searchKeys: ['gene', 'hgnc_id', 'diseases', 'top_acmg'] });

    // ── Gene × Variant table ──
    const varCols = [
      { key: 'gene', label: 'Gene', render: r => `<strong style="color:var(--accent)">${esc(r.gene)}</strong>` },
      { key: 'hgnc_id', label: 'HGNC', render: r => `<span style="color:var(--muted);font-size:11px">${esc(r.hgnc_id)}</span>` },
      { key: 'disease', label: 'Disease', render: r => `<span>${esc(r.disease)}</span>` },
      { key: 'acmg', label: 'ACMG', render: r => badge(r.acmg, ACMG_BADGE) },
      { key: 'status', label: 'Status', render: r => badge(r.status, STATUS_BADGE) },
      { key: 'hgvs', label: 'HGVS (c.)', render: r => `<span style="font-size:11px;color:var(--muted)">${esc(r.hgvs)}</span>` },
      { key: 'allelic', label: 'Zygosity', render: r => `<span style="font-size:11px">${esc(r.allelic)}</span>` },
    ];
    renderTable('tbl-genes', g.rows, varCols,
      { defaultSort: 'gene', defaultDir: 'asc', searchKeys: ['gene', 'disease', 'hgvs', 'acmg'] });

    // ── Explorer ──
    const rebuild = () => buildExplorer();
    ['gd-view', 'gd-top', 'gd-metric'].forEach(id =>
      document.getElementById(id).addEventListener('change', rebuild));
    buildExplorer();
  },
};

async function buildExplorer() {
  const view = document.getElementById('gd-view')?.value || 'gene-per-dx';
  const topN = parseInt(document.getElementById('gd-top')?.value || '15');
  const metric = document.getElementById('gd-metric')?.value || 'patients';

  const chartEl = document.getElementById('chart-gene-dx');
  const heatmapEl = document.getElementById('gd-heatmap');
  const phEl = document.getElementById('gd-placeholder');
  const legendEl = document.getElementById('gd-legend');

  const data = await callApi('gene_dx_explorer', { view, top_n: topN, metric });

  if (data.empty) {
    phEl.textContent = 'No gene data available.';
    phEl.style.display = 'block';
    chartEl.style.display = 'none';
    heatmapEl.style.display = 'none';
    return;
  }

  if (view === 'heatmap') {
    destroyChart('gene-dx');
    chartEl.style.display = 'none';
    heatmapEl.style.display = 'block';
    phEl.style.display = 'none';
    legendEl.innerHTML = '';

    const { genes, diagnoses, matrix, max } = data;
    const CELL = 36, LPAD = 120, TPAD = 110, CPAD = 8;
    const W = LPAD + diagnoses.length * CELL + CPAD;
    const H = TPAD + genes.length * CELL + CPAD;

    const cells = genes.flatMap((gName, gi) => diagnoses.map((dName, di) => {
      const v = matrix[gi][di];
      if (!v) return '';
      const intensity = v / max;
      const bg = `rgba(37,99,235,${(0.08 + intensity * 0.85).toFixed(2)})`;
      const tx = intensity > 0.55 ? '#fff' : '#1a2332';
      const x = LPAD + di * CELL, y = TPAD + gi * CELL;
      return `<rect x="${x}" y="${y}" width="${CELL - 2}" height="${CELL - 2}" rx="3" fill="${bg}"/>
              <text x="${x + CELL / 2 - 1}" y="${y + CELL / 2 + 4}" text-anchor="middle" font-size="10"
                fill="${tx}" font-family="${NUM_FONT}">${v}</text>`;
    })).join('');

    const gLabels = genes.map((gName, i) =>
      `<text x="${LPAD - 6}" y="${TPAD + i * CELL + CELL / 2 + 4}" text-anchor="end" font-size="11"
        fill="#334155" font-family="${NUM_FONT}">${esc(gName)}</text>`).join('');

    const dLabels = diagnoses.map((dName, i) => {
      const label = dName.length > 18 ? dName.slice(0, 17) + '…' : dName;
      const x = LPAD + i * CELL + CELL / 2;
      return `<text x="${x}" y="${TPAD - 6}" text-anchor="start" font-size="10"
        fill="#334155" font-family="${NUM_FONT}"
        transform="rotate(-45,${x},${TPAD - 6})">${esc(label)}</text>`;
    }).join('');

    heatmapEl.innerHTML = `
      <svg width="${W}" height="${H}" style="min-width:${W}px">${gLabels}${dLabels}${cells}</svg>
      <div style="font-size:10px;color:var(--muted);margin-top:6px">
        Colour intensity = ${metric === 'patients' ? 'unique patients' : 'total records'} · white = 0
      </div>`;
    return;
  }

  // ── Stacked bars ──
  heatmapEl.style.display = 'none';
  phEl.style.display = 'none';
  chartEl.style.display = 'block';

  const primaryKeys = data.primary_keys;
  const datasets = data.datasets.map((ds, si) => ({
    label: ds.label,
    data: ds.data,
    backgroundColor: GD_PALETTE[si % GD_PALETTE.length],
    borderWidth: 0,
  }));

  const barH = Math.max(28, Math.min(50, 700 / primaryKeys.length));
  chartEl.parentElement.style.height = Math.max(300, primaryKeys.length * barH + 60) + 'px';

  destroyChart('gene-dx');
  charts['gene-dx'] = new Chart(chartEl, {
    type: 'bar',
    data: { labels: primaryKeys, datasets },
    options: {
      indexAxis: 'y',
      responsive: true, maintainAspectRatio: false,
      plugins: {
        legend: { display: false },
        tooltip: { callbacks: { label: ctx => `${ctx.dataset.label}: ${ctx.raw.toLocaleString()} ${metric}` } },
      },
      scales: {
        x: {
          stacked: true, ticks: { color: '#94a3b8', font: { size: 10, family: NUM_FONT } }, grid: { color: 'rgba(0,0,0,.06)' },
          title: { display: true, text: metric === 'patients' ? 'Unique patients' : 'Records', color: '#7a8694', font: { size: 10, family: NUM_FONT } },
        },
        y: { stacked: true, ticks: { color: '#334155', font: { size: 10, family: NUM_FONT } }, grid: { display: false } },
      },
    },
  });

  legendEl.innerHTML = datasets.slice(0, 40).map((ds, si) => `
    <span style="display:inline-flex;align-items:center;gap:4px;white-space:nowrap">
      <span style="width:10px;height:10px;border-radius:2px;background:${GD_PALETTE[si % GD_PALETTE.length]};flex-shrink:0"></span>
      <span style="font-family:var(--font-num);font-size:10px">${esc(ds.label.length > 30 ? ds.label.slice(0, 29) + '…' : ds.label)}</span>
    </span>`).join('') +
    (datasets.length > 40 ? `<span style="color:var(--muted);font-size:10px">+${datasets.length - 40} more</span>` : '');
}
